File: leetcode/hard/patterns/sliding-window/76. Minimum Window Substring.py
def isValid(freq_s, freq_t):
    for ch in freq_t:
        if freq_s.get(ch,0) < freq_t[ch]:
            return False
    return True

def minWindow(s: str, t: str) -> str:
    low = 0
    high = 0
    freq_s = {}
    freq_t = {}
    res = ""
    i = 0
    while i < len(t):
        freq_t[t[i]] = freq_t.get(t[i], 0) + 1
        i += 1
    while high < len(s):
        freq_s[s[high]] = freq_s.get(s[high], 0) + 1

        while isValid(freq_s=freq_s, freq_t=freq_t):
            curr = s[low:high+1]
            if  res == "" or len(curr) < len(res):
                res = curr
            freq_s[s[low]] -= 1
            if freq_s[s[low]] == 0:
                del freq_s[s[low]]
            low += 1
        high += 1
    return res

print(minWindow(s="ADOBECODEBANC", t="ABC"))

File: revision/sliding-window/76.Minimum-Window-Substring.cpp
#include<iostream>
#include<vector>
#include<string>
#include<unordered_map>
using namespace std;
class Solution {
public:
    bool isvalid(unordered_map<char, int> &fs, unordered_map<char,int> &ft) {
        for (auto x: ft) {
            if(fs[x.first] < x.second) {
                return false;
            }
        }
        return true;
    }
    string minWindow(string &s, string &t) {
        int high = 0;
        int low = 0;
        int res = INT_MAX;
        int start = 0;
        int i = 0;
        unordered_map<char,int> fs;
        unordered_map<char,int> ft;
        while ( i < t.size()) {
            ft[t[i]]++;
            i++;
        }
        while (high < s.size()) {
            fs[s[high]]++;

            while(isvalid(fs,ft)) {
                int window_len = (high-low) +1;
                if(window_len < res) {
                    res = window_len;
                    start = low;
                }
                fs[s[low]]--;
                if(fs[s[low]] == 0) {
                    fs.erase(s[low]);
                }
                low++;

            }
            high++;
        }
        if (res == INT_MAX) {
            return "";
        }
        return s.substr(start, res);
    }
};
int main() {
    Solution sol;
    string s = "ADOBECODEBANC";
    string t = "ABC";
    string ans = sol.minWindow(s,t);
    cout << ans;
    return 0;



}